toolchain/check/impl_lookup.cpp

```diff
@@ -492,11 +490,18 @@ struct CandidateImpl {
 
 // Returns the list of candidates impls for lookup to select from.
 static auto CollectCandidateImplsForQuery(
-    Context& context, const TypeStructure& query_type_structure,
+    Context& context, bool final_only,
+    const TypeStructure& query_type_structure,
     SemIR::SpecificInterface& query_specific_interface)
     -> llvm::SmallVector<CandidateImpl> {
   llvm::SmallVector<CandidateImpl> candidate_impls;
   for (auto [id, impl] : context.impls().enumerate()) {
+    if (final_only) {
+      if (!IsImplEffectivelyFinal(context, impl)) {
+        continue;
+      }
+    }
```

Comment:
Optional: combine these
```suggestion
    if (final_only && !IsImplEffectivelyFinal(context, impl)) {
      continue;
    }
```